consensus/core/src/dag_state.rs

```diff
@@ -27,48 +28,65 @@ const BLOCK_CACHED_ROUNDS: Round = 100;
-    // last committed leader round.
+    // Caches blocks within CACHED_ROUNDS from the last committed round per authority.
+    // Note: uncommitted blocks will always be in memory.
     recent_blocks: BTreeMap<BlockRef, VerifiedBlock>,
 
-    // All accepted blocks have their refs cached. Cached refs are never removed for now.
-    // Each element in the vector contains refs for the authority corresponding to its index.
+    // Accepted blocks have their refs cached. Cached refs are never removed until restart.
+    // Each element in the Vec corresponds to the authority with the index.
     cached_refs: Vec<BTreeSet<BlockRef>>,
 
+    // Last consensus commit of the dag.
+    last_commit: Option<Commit>,
+
     // Persistent storage for blocks, commits and other consensus data.
     store: Arc<dyn Store>,
 }
 
 #[allow(unused)]
 impl DagState {
-    pub(crate) fn new(
-        context: Arc<Context>,
-        blocks: Vec<VerifiedBlock>,
-        store: Arc<dyn Store>,
-    ) -> Self {
+    /// Initializes DagState from storage.
+    pub(crate) fn new(context: Arc<Context>, store: Arc<dyn Store>) -> Self {
         let num_authorities = context.committee.size();
+        let last_commit = store.read_last_commit().unwrap();
+        let last_committed_rounds = match &last_commit {
+            Some(commit) => commit.last_committed_rounds.clone(),
+            None => vec![0; num_authorities],
+        };
+
         let mut state = Self {
             context,
             recent_blocks: BTreeMap::new(),
             cached_refs: vec![BTreeSet::new(); num_authorities],
+            last_commit,
             store,
         };
 
-        for block in blocks {
-            state.add_block(block);
+        for (i, round) in last_committed_rounds.into_iter().enumerate() {
+            let authority_index = state.context.committee.to_authority_index(i).unwrap();
+            let blocks = state
+                .store
+                .scan_blocks_by_author(authority_index, round.saturating_sub(CACHED_ROUNDS))
+                .unwrap();
+            for block in blocks {
+                state.accept_block(block);
+            }
         }
 
         state
     }
 
-    pub(crate) fn add_block(&mut self, block: VerifiedBlock) {
+    /// Accepts a block into DagState and keeps it in memory.
+    pub(crate) fn accept_block(&mut self, block: VerifiedBlock) {
         let block_ref = block.reference();
         self.recent_blocks.insert(block_ref, block);
         self.cached_refs[block_ref.author].insert(block_ref);
     }
 
-    pub(crate) fn get_block(&self, _reference: BlockRef) -> Option<VerifiedBlock> {
-        unimplemented!()
+    /// Gets a copy of an uncommitted block. Returns None if not found.
+    /// Uncommitted must be in memory, so only in-memory blocks are checked.
+    pub(crate) fn get_uncommitted_block(&self, reference: &BlockRef) -> Option<VerifiedBlock> {
```

Comment:
Is it possible to only have up to `CACHED_ROUNDS` worth of blocks in memory but have the block requested be outside of those rounds and in store? If so should we fall back to reading from store?

Reply:
I may have worded it poorly, but the scenario I was thinking about was during catchup when we may have more than 100 rounds of uncommitted blocks. In this case if there is a call to get_uncommitted_block that is outside of that range this would return None. But I see now that this won't happen and we will always be able to use the blocks in memory.